File: android/src/main/kotlin/com/skaiwalk/sifli_ezip/SifliEzipPlugin.kt
package com.skaiwalk.sifli_ezip

import com.sifli.ezip.sifliEzipUtil
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** SifliEzipPlugin */
class SifliEzipPlugin :
    FlutterPlugin,
    MethodCallHandler {
    // The MethodChannel that will the communication between Flutter and native Android
    //
    // This local reference serves to register the plugin with the Flutter Engine and unregister it
    // when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "sifli_ezip")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "pngToEzip" -> {
                try {
                    val pngData = call.argument<ByteArray>("pngData")
                    val colorType = call.argument<String>("colorType")
                    val ezipColorType = call.argument<Int>("ezipColorType")
                    val ezipBinType = call.argument<Int>("ezipBinType")
                    val boardType = call.argument<Int>("boardType")

                    if (pngData == null || colorType == null || ezipColorType == null ||
                        ezipBinType == null || boardType == null) {
                        result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
                        return
                    }

                    val ezipData = sifliEzipUtil.pngToEzip(
                        pngData,
                        colorType,
                        ezipColorType,
                        ezipBinType,
                        boardType
                    )

                    if (ezipData != null) {
                        result.success(ezipData)
                    } else {
                        result.error("CONVERSION_FAILED", "Failed to convert PNG to EZIP", null)
                    }
                } catch (e: Exception) {
                    result.error("EXCEPTION", e.message, null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/SifliEzipPlugin.swift
import Flutter
import UIKit
import eZIPSDK

public class SifliEzipPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "sifli_ezip", binaryMessenger: registrar.messenger())
    let instance = SifliEzipPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "pngToEzip":
      guard let args = call.arguments as? [String: Any],
            let pngData = args["pngData"] as? FlutterStandardTypedData,
            let colorType = args["colorType"] as? String,
            let ezipColorType = args["ezipColorType"] as? Int,
            let ezipBinType = args["ezipBinType"] as? Int,
            let boardType = args["boardType"] as? Int else {
        result(FlutterError(code: "INVALID_ARGUMENTS",
                           message: "Missing required arguments",
                           details: nil))
        return
      }

      // Call Sifli eZIPSDK
      if let ezipResult = ImageConvertor.eBin(
        fromPNGData: pngData.data,
        eColor: colorType,
        eType: UInt8(ezipColorType),
        binType: UInt8(ezipBinType),
        boardType: SFBoardType(rawValue: UInt(boardType)) ?? SFBoardType(rawValue: 0)!
      ) {
        result(FlutterStandardTypedData(bytes: ezipResult))
      } else {
        result(FlutterError(code: "CONVERSION_FAILED",
                           message: "Failed to convert PNG to EZIP",
                           details: nil))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
